toolchain/check/deduce.cpp

```diff
@@ -0,0 +1,196 @@
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/check/deduce.h"
+
+#include "toolchain/base/kind_switch.h"
+#include "toolchain/check/context.h"
+#include "toolchain/check/generic.h"
+#include "toolchain/check/subst.h"
+#include "toolchain/sem_ir/typed_insts.h"
+
+namespace Carbon::Check {
+
+namespace {
+struct DeductionWorklist {
+  auto Add(SemIR::InstId param, SemIR::InstId arg, bool needs_substitution)
+      -> void {
+    deductions.push_back(
+        {.param = param, .arg = arg, .needs_substitution = needs_substitution});
+  }
+
+  auto AddBlock(llvm::ArrayRef<SemIR::InstId> params,
+                llvm::ArrayRef<SemIR::InstId> args, bool needs_substitution)
+      -> void {
+    if (params.size() != args.size()) {
+      return;
+    }
+    for (auto [param, arg] : llvm::zip_equal(params, args)) {
+      Add(param, arg, needs_substitution);
+    }
+  }
+
+  auto AddBlock(SemIR::InstBlockId params, llvm::ArrayRef<SemIR::InstId> args,
+                bool needs_substitution) -> void {
+    AddBlock(context.inst_blocks().Get(params), args, needs_substitution);
+  }
+
+  auto AddBlock(SemIR::InstBlockId params, SemIR::InstBlockId args,
+                bool needs_substitution) -> void {
+    AddBlock(context.inst_blocks().Get(params), context.inst_blocks().Get(args),
+             needs_substitution);
+  }
+
+  struct PendingDeduction {
+    SemIR::InstId param;
+    SemIR::InstId arg;
+    bool needs_substitution;
+  };
+  Context& context;
+  llvm::SmallVector<PendingDeduction> deductions;
+};
+}  // namespace
+
+static auto NoteGenericHere(Context& context, SemIR::GenericId generic_id,
+                            Context::DiagnosticBuilder& diag) -> void {
+  CARBON_DIAGNOSTIC(DeductionGenericHere, Note,
+                    "While deducing parameters of generic declared here.");
+  diag.Note(context.generics().Get(generic_id).decl_id, DeductionGenericHere);
+}
+
+auto DeduceGenericCallArguments(Context& context, Parse::NodeId node_id,
+                                SemIR::GenericId generic_id,
+                                SemIR::SpecificId enclosing_specific_id,
+                                SemIR::InstBlockId implicit_params_id,
+                                SemIR::InstBlockId params_id,
+                                SemIR::InstId self_id,
+                                llvm::ArrayRef<SemIR::InstId> arg_ids)
+    -> SemIR::SpecificId {
+  DeductionWorklist worklist = {.context = context};
+
+  // TODO: Perform deduction for type of self
+  static_cast<void>(implicit_params_id);
+  static_cast<void>(self_id);
+
+  // Copy any outer generic arguments from the specified instance and prepare to
+  // substitute them into the function declaration.
+  llvm::SmallVector<SemIR::InstId> results;
+  llvm::SmallVector<Substitution> substitutions;
+  if (enclosing_specific_id.is_valid()) {
```

Comment:
All our `Id` types have an `::Invalid` value that represents "absent", as if there's an implicit `optional` wrapping them all. We use the invalid state to mean "none" pretty heavily.

(For what it's worth, I'm not sure "valid" / "invalid" is the best terminology here, because it has nothing to do with program validity, and we have a separate `Error` value for some of our `Id` types as a placeholder for "couldn't form this value due to an error". Might be worth changing this to `::Absent` and `is_present()` or similar.)